### PR TITLE
Incremental 03a

### PR DESCRIPTION
**Pull request by rgayatri23**
Unit Test for parallel_for functor

### CODE REVIEW

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+#include <cstdio>
+#include <sstream>
+#include <type_traits>
+#include <gtest/gtest.h>
+
+/// @Kokkos_Feature_Level_Required:3
+
+namespace Test {
+
+using dataType = double;
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: `s/dataType/DataType/`

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+// parallel-for unit test
+// create an array of double dataType and add a constatnt to all elements of the
+// array in a parallel-for
+
+template <class ExecSpace>
+struct array {
+  const int num_elements = 10;
+  const double value     = 0.5;
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]:
```suggestion
  const dataType value     = 0.5;
```

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 3]
```diff
@@ -0,0 +1,156 @@
...
+#include <gtest/gtest.h>
+
+/// @Kokkos_Feature_Level_Required:3
+
+namespace Test {
+
+using dataType = double;
+
+// parallel-for unit test
+// create an array of double dataType and add a constatnt to all elements of the
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]:
```suggestion
// create an array of double dataType and add a constant to all elements of the
```

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+// array in a parallel-for
+
+template <class ExecSpace>
+struct array {
+  const int num_elements = 10;
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: Never used.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+  array(dataType *data) : _data(data) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const int i) const { _data[i] = i * value; }
+};
+
+template <class ExecSpace>
+struct TestParallel_For {
+  int num_elements     = 10;
+  const dataType value = 0.5;
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: Must match `array::value`.  This is not a good idea.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 6]
```diff
@@ -0,0 +1,156 @@
...
+  dataType *DeviceData, *HostData;
+
+  // memory_space for the memory allocation
+  typedef typename TEST_EXECSPACE::memory_space MemSpaceD;
+  typedef Kokkos::HostSpace MemSpaceH;
+
+  int compare_equal(dataType *data) {
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: If you are checking the answer on the host, you might as well call the googletest assertion macros.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp (hunk 3 above)
Comment (dalg24) [reply core/unit_test/incremental/Test03a_ParallelFor.hpp]: Also this is not accurate.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 7]
```diff
@@ -0,0 +1,156 @@
...
+
+    // Allocate Memory for both device and host memory spaces
+    DeviceData = (dataType *)Kokkos::kokkos_malloc<MemSpaceD>(
+        "dataD", num_elements * sizeof(dataType));
+    HostData = (dataType *)Kokkos::kokkos_malloc<MemSpaceH>(
+        "dataH", num_elements * sizeof(dataType));
+
+    // parallel-for functor called for num_elements elements
+    Kokkos::parallel_for("RangePolicy_ParallelFor",
+                         range_policy(0, num_elements),
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: All code but this line is duplicated from `simple_test()`.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 8]
```diff
@@ -0,0 +1,156 @@
...
+namespace Test {
+
+using dataType = double;
+
+// parallel-for unit test
+// create an array of double dataType and add a constatnt to all elements of the
+// array in a parallel-for
+
+template <class ExecSpace>
+struct array {
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: Not a good name.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 9]
```diff
@@ -0,0 +1,156 @@
...
+  const int num_elements = 10;
+  const double value     = 0.5;
+  dataType *_data;
+  typedef typename TEST_EXECSPACE::memory_space memory_space;
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: Never used.  Class does not need to be templated if you drop it.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 10]
```diff
@@ -0,0 +1,156 @@
...
+
+namespace Test {
+
+using dataType = double;
+
+// parallel-for unit test
+// create an array of double dataType and add a constatnt to all elements of the
+// array in a parallel-for
+
+template <class ExecSpace>
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: Template argument is not used.

**Comment by rgayatri23** [comment]
Retest/Re-review this please.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+// Questions? Contact Christian R. Trott (crtrott@sandia.gov)
+//
+// ************************************************************************
+//@HEADER
+*/
+
+#include <Kokkos_Core.hpp>
+#include <cstdio>
+#include <sstream>
+#include <type_traits>
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: Do not include headers that are not needed.  I don't think you currently use anything from `<cstdio>`, `<sstream>`, not `<type_traits>`.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+  Functor(DataType *data) : _data(data) {}
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const int i) const { _data[i] = i * value; }
+};
+
+template <class ExecSpace>
+struct TestParallel_For {
+  int num_elements = 10;
+  DataType *DeviceData, *HostData;
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: Try not to mix styles.  I don't know of any other pace in Kokkos that uses PascalCase for variables.

File: core/unit_test/incremental/Test03a_ParallelFor.hpp [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+
+    // Free the allocated memory
+    Kokkos::kokkos_free<MemSpaceD>(DeviceData);
+    Kokkos::kokkos_free<MemSpaceH>(HostData);
+  }
+
+  // A parallel_for test with user defined RangePolicy
+  void range_policy() {
+    // Create a range policy for the parallel_for
+#if defined(KOKKOS_ENABLE_CUDA)
```
Comment (dalg24) [inline core/unit_test/incremental/Test03a_ParallelFor.hpp]: Why do you need the distinction?

Comment (rgayatri23) [reply core/unit_test/incremental/Test03a_ParallelFor.hpp]: I just kept that to test the schedule type feature. Not doing anything specific with it.

**Comment by crtrott** [comment]
closed in favour of #2633 Note that this pull request would have deleted existing tests ...

**Pull request closed without merging**